### DISCUSSION TITLE
integration with react-native-web

### DISCUSSION THREAD

**Discussion by mannbonn** [invertase/react-native-firebase#4417] [Q&A] [answered]
[steps to reproduce]
1) Follow this https://medium.com/@aureliomerenda/create-a-native-web-app-with-react-native-web-419acac86b82
2) When the guide does say **`Remove react-native example components`** just paste this code to replace the whole _./src/App.js_
```
import React from 'react';
import {View, Text} from 'react-native';

export default function App() {
  return(
      <View> <Text>{"Hello world"}</Text> </View>
  )
}
```

3) Keep following the guide.
At the end of this steps alone, a webpage open and the react native code will be shown app in it. Neat. 
4) install '@react-native-firebase/app' and '@react-native-firebase/auth'
5) add `import auth from '@react-native-firebase/auth'` to the App.js
6) run the web scripts again, watch it fail. 

The console error message is completely bogus, as the file are in the place it says it could not find them. 
chokidar complains of a file that can't be written in C:.
The browser console says Uncaught SyntaxError: expected expression, got '!'

I'm ready to add anything that could help pinpointing the problem, as to the top of my knowledge there is not much I can do. Sorry if the question is missing any information, but I'm just overflown by this problem.

**Comment by mikehardy** [comment, marked as answer]
We simply do not support react-native-web and no one has proposed themselves to step up to contribute the PRs such that we *do* support react-native-web (and most important!) to maintain it going forward.

What I do (using reactxp, which also supports web, similar to react-native-web) is that when I do the web build I cut react-native-firebase out completely and I pull in firebase-js-sdk directly from the google repos. No attempt to build react-native-firebase is made in my web build, because it doesn't work

With no intention at being discouraging but to set my personal boundaries clearly - I will make no attempt to support react-native-web here though I'd like to see it of course. We need someone from the community to step up and commit to maintaining it for it to happen

**Reply by Phazoning** [reply]
Just a small question, do you know how to use crashlytics with the sdk? I've been trying for a while with no success whatsoever

**Reply by mikehardy** [reply]
"with the sdk" - which SDK? firebase-js-sdk does not have crashlytics, so the answer is just "you can't". If you mean android and ios, then the way to use it in react-native is https://rnfirebase.io/crashlytics/usage

**Reply by Phazoning** [reply]
I thought that was clear from the context, sorry, and yes, the JS one. It's not as much of a matter of using it with android/ios, but with docz, because I've applied the fix as they wrote on their GitHub (basically, usring react-native-web instead and telling gatsby to use it whenever react-native is used) and this is the single thing I cannot translate, so to say, to web, because sure, the sdk fixes most of the errors, but I still need crashlytics. Thanks anyway

**Reply by josefnorlin** [reply]
Thank you for clarifying this. I ran into the same issues albeit with other (probably) bogus error messages trying to run "npx expo start --web" on my managed expo build (version 47.0.12). 

But mine read (for others maybe googling this): 

> Uncaught Error: You attempted to use a firebase module that's not installed on your Android project by calling firebase.app().
> Ensure you have:
> 1) imported the 'io.invertase.firebase.app.ReactNativeFirebaseAppPackage' module in your 'MainApplication.java' file.
> 2) Added the 'new ReactNativeFirebaseAppPackage()' line inside of the RN 'getPackages()' method list.
> See http://invertase.link/android for full setup instructions.`
> 

But I wonder if anything regarding this has changed during this year and a half?

And if not. Do you know how I can help? I think this would come in really handy since Firebase + React Native would be a killer solution for developing to ios, android, and web at once. Or should I just give up and create a separate web project and use firebase-js-sdk?

**Reply by cglacet** [reply]
> We simply do not support react-native-web

@mikehardy Is this still true today? If not, would this be the proper replacement: [firebase/auth + firebase/app](https://firebase.google.com/docs/auth/web/start) (I'm using Nextjs for the web version of our app, not sure if it makes any difference)?

Or maybe this: [reactfire](https://github.com/FirebaseExtended/reactfire#reactfire)?

**Reply by mikehardy** [reply]
We still do not support react-native-web. It is on our roadmap (after adopting the modular API here, in progress #6705)
The best way to support it that I know of is by doing shims like I demonstrate in the aging but worked https://github.com/invertase/react-native-firebase-authentication-example template - specifically it used craco to override babel configs such that when babel was in play (react-native-web...) attempts to import the `@react-native-firebase/etc` packages were overridden to shims, and the shims loaded the firebase-js-sdk package and smoothed over any API differences (there weren't many, the shims were tiny)

**Reply by BoavistaLudwig** [reply]
@mikehardy ,
thanks for this hint to get React Native for Web working with Firebase RN SDK.

I already had RN for Web working with Firebase Web SDK, see here: www.heybossapp.com
But, I need Dynamic Links to create a func. that lets users invite other users to their "missions"... 🤯

I am still newish to web/mobile dev. and must ask this in advance, b/c otherwise, I just go for it and burn time:
Do I only have to add a couple of "shims" as I see here (https://github.com/invertase/react-native-firebase-authentication-example/tree/main/template/src/shims) to get Firebase RN SDK working on the web? Doesnt sound too bad. I am using Expo though, and probably wont need craco...? 

Thanks for the great work!

**Reply by mikehardy** [reply]
Unfortunately I have no idea how to get Expo configured to work with different modules loading on web vs native - I explicitly avoided Expo in the template because a) I wanted a smaller / more stock template and Expo is kinda huge and b) I figured Expo had already figured that stuff out and documented it so they didn't need the help

Which...is not really helpful to you, apologies. *If* Expo allows you to alias modules when loading them depending on platform (such that you can load your shims for web) *then* yes you just add some web shims and use them to load the firebase-js-sdk imports then re-export them like they were the react-native-firebase style. That's the whole "big idea" in the template, conceptually. If there is any way to do anything like that in Expo, it should be possible to make it work for you

**Reply by cglacet** [reply]
Maybe this? [customizing-webpack](https://docs.expo.dev/guides/customizing-webpack/) using [aliases](https://webpack.js.org/configuration/resolve/#resolvealias)

**Reply by mikehardy** [reply]
That looks right on target

**Reply by cglacet** [reply]
By the way, I didn't answered earlier but your suggestion worked great for me (on a nextjs project).

**Reply by BoavistaLudwig** [reply]
So, apparently with Expo, 
I only had to add shims files to mix Firebase Web SDK & React Native Firebase, 
and I followed @mikehardy 's guidance with his demo app at https://github.com/invertase/react-native-firebase-authentication-example/tree/main/template/src/shims. :)

...problem solved 🥳, tackling the next ;-)

**Reply by vipulbhj** [reply]
@BoavistaLudwig Hello good sir, could you explain how you solved this ? I am also using expo and just saw it it now working on Web. I saw your above comment, but now sure of the exact steps needed for this ? 

Can you share that maybe ?

**Reply by mikehardy** [reply]
Just a note that this is outdated. 

react-native-firebase supports the web now (for firebase modules that have web support!) 

https://rnfirebase.io/platforms#other-platforms

**Comment by BoavistaLudwig** [comment]
@mikehardy working with shim files and your template, 


```
import firebase from 'firebase/compat/app';
import 'firebase/compat/auth';

import initializeApp from './firebase-init';
initializeApp();

const auth = firebase.auth;

// Want to do local development?
// Uncomment this and use `yarn test:emulator:start`
// auth().useEmulator('http://localhost:9099');

export default auth;

```
at https://github.com/invertase/react-native-firebase-authentication-example/blob/main/template/src/shims/firebase-auth-web.ts


how do you add 

```
import {
  getReactNativePersistence,
  initializeAuth,
} from "firebase/auth/react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
```

to achieve 

```
export const auth = initializeAuth(app, {
  persistence: getReactNativePersistence(AsyncStorage),
});
```

??

Because Expo SDK 48 doesn't come with Async Storage any more, authentication on React Native doesn't persist out of the box. I had to add these additional imports before I tackled the shim files scenario, and it worked. 

Right now, I can't get my head around it when using shims. Web persistence already works out of the box...

**Reply by mikehardy** [reply]
I believe this is handled now in our new "other" (read as: fallback to firebase-js-sdk when not on ios/android) support

https://rnfirebase.io/platforms#other-platforms

**Comment by adminsisnet** [comment]
Could you see if the  Machine learning abiliy in my logic can help you and validate my 70% GPU thermal efficacy claim?https://github.com/adminsisnet/0-0logicDuplicate
